music21/stream/base.py

```diff
@@ -3709,7 +3709,71 @@ def getElementsByOffset(
             sIterator = sIterator.getElementsByClass(classList)
         return sIterator
 
-    def getElementAtOrBefore(self, offset, classList=None):
+    def getContextByClass(
+        self,
+        className,
+        *,
+        getElementMethod='getElementAtOrBefore',
+        sortByCreationTime=False,
+        followDerivation=True,
+        priorityTargetOnly=False,
```

Comment:
can we make the new behavior optional with a keyword like `includeInternalContexts=SearchInternalContexts.RECURSE_ZERO` where SearchInternalContexts is an enum of `NEVER, AT_ZERO, and ALL, RECURSE_ZERO, and RECURSE_ALL`

I think that it would be strange that a Measure object getContextByClass('TimeSignature') will find it, but that Part object would not.  So RECURSE or not would determine what iterator to call, and ZERO vs ALL would put in the limitation of getElementsByOffsetInHierarchy at zero or not.

Reply:
Wow, yes, I can see the utility of this. That will make this probably the most powerful method in m21, then--to be able to in one method call rip up, down, left, or right through the containers. I think this is manageable, will try to get to it in the next couple weeks. Once this is in I already have a followup patch prepared for the deprecations that depend on this change.

Reply:
Despite video games not being a meaningful part of my life at any point, I've noticed we'll be creating music21's joystick controller.

Reply:
ha-ha-ha-ha!!!!

I think we should probably use one of the fancy new StringEnums to replace 'getElementAtOrBefore' etc. -- having routines that search for 'At' in the string and do different things based on it probably doesn't make sense...

Reply:
Shout out to Chris Ariza who initially implemented this when I decided I wanted it, and somehow figured it out how to get it done.  And Josiah Oberholtzer who figured out how to speed up the code immensely for large scores.

Reply:
After thinking about this for a bit I'm having concerns about API surface area.

I think "ALL" is what would have solved the original problem with finding the clef at 0 (or midmeasure, at 2, etc.), and that can be had with either "ALL" or "gEBC" or whatnot:

**on main branch, demoing that we just need to document how to not limit by offsets:**
```python
>>> s = stream.Stream()
>>> s.insert(clef.BassClef())
>>> s.getContextByClass(clef.Clef) is None
True
>>> s.getContextByClass(clef.Clef, getElementMethod='dont_mind_me_just_a_synonym_for_all_or_gEBC')
<music21.clef.BassClef>
>>> s.getContextByClass(clef.Clef, getElementMethod='getElementsAtOrAfter')
<music21.clef.BassClef>
```

Then the other piece of this was whether to provide syntactic sugar to let people run two searches -- first, internal substreams, then context streams -- but then wouldn't it be possible a user would want different at/after/before for the two searches? Any clef in measure 20 including midmeasure clefs, so atOrAfter 0.0, but finding none, want any clef before 80.0, where that measure happens to sit in the context stream, which would be AtOrBefore. And so on--leading to me believe I shouldn't be adding this after all.

Really I think the original bug was just stumbling over the fact that in a tree search, I can't search other nodes at the same offset. If I happen to be at 0.0 but before the other two elements at 0.0, and I'm searching at-or-after: out of luck.

Reply:
Works equally well, and is used in the package, but just wasn't documented:

```python
>>> s.getContextByClass(clef.Clef, getElementMethod='getElementAtOrBeforeOffset')
<music21.clef.BassClef>
```